refactor: move interactions with meta into dedicated module

anonymous123-code opened the pull request:
This PR moves all interactions with meta into one file, which allows for easier improvements in the future and reduces code duplication.
It additionally adds a config option to specify the meta base url.
It might be a good idea to either simplify `get_component_meta` (Which matches the old `Component::into_meta`) or add similar file-based storage to `get_component_index` in order to improve consistency.

anonymous123-code committed:
Meta refactor: move into dedicated struct

kb-1000 commented on helixlauncher-core/src/launch/prepared.rs:
```diff
@@ -317,21 +313,13 @@ pub async fn prepare_launch(
+    for item in HelixLauncherMeta::new(config)
+        .get_component_index(&path)
+        .await
+        .expect("Error while checking if version exists")
```
I don't think this is currently used anywhere, but panicking on non-logic errors is bad... not really sure if it _will_ ever be used though (if it is, it'd go in the meta module)

anonymous123-code replied:
This was the old behavior so I kept it. I can do some more refactoring if you want, too

kb-1000 commented on helixlauncher-core/src/meta.rs:
```diff
@@ -0,0 +1,91 @@
+impl<'a> HelixLauncherMeta<'a> {
+    pub fn new(config: &'a Config) -> HelixLauncherMeta<'a> {
+        HelixLauncherMeta {
+            client: Client::new(),
```
Creating a new Client instance on each invocation is bad, though if the Meta instance is kept around that would probably be fine

anonymous123-code replied:
Well alternatively the client would need to be kept around (or newly instanciated each time) so this shouldnt make much of a difference

kb-1000 commented on helixlauncher-core/src/meta.rs:
```diff
@@ -0,0 +1,91 @@
+    #[error(transparent)]
+    ParseError(#[from] serde_json::Error),
+    #[error(transparent)]
+    ReqwestError(#[from] reqwest::Error),
```
This may need more variants, if the component doesn't exist it should immediately say so (though this is probably fine for errors other than not found)

kb-1000 commented on helixlauncher-core/src/meta.rs:
```diff
@@ -0,0 +1,91 @@
+
+use crate::config::Config;
+
+pub struct HelixLauncherMeta<'a> {
```
I don't really like the name here, honestly... `helixlauncher_core::meta::HelixLauncherMeta`
Something like `MetaClient` or `ComponentMetaClient` could work?

kb-1000 reviewed (changes requested):
I'd also like to note that `format!()` can accept local variables inside the `{}` in the format string which IMO can make things more readable

Ecorous commented on helixlauncher-core/src/config.rs:
```diff
@@ -29,6 +30,9 @@ pub struct Config {
     assets_dir: PathBuf,
+
+    #[serde(default = "meta_url_default")]
+    meta_url: String,
```
I don't. The default should always be listed in the config

anonymous123-code replied:
In the future, it might be possible (and not *too* difficult) to make it so that multiple component sources are included at once. (Similarly to how `get_component_meta` already falls back to a local version) But thats a discussion for another place. For the time being I agree with Ecorous

kb-1000 replied:
This especially is the kind of default that _can_ change, and isn't too unlikely to change. In fact, I think it may make sense to do so soon, as currently we have no version in the URL.
If the user didn't touch this sort of setting, then why would the old default stay? That seems like fairly unexpected behavior to me.
Falling back to a local version as it's currently implemented is a crude form of caching with offline mode support, and at least by me not intended to be used to add custom component jsons. That should be instance level just like with MultiMC, otherwise you'd have two parts to copy around.

Ecorous replied:
> This especially is the kind of default that _can_ change, and isn't too unlikely to change. In fact, I think it may make sense to do so soon, as currently we have no version in the URL. If the user didn't touch this sort of setting, then why would the old default stay? That seems like fairly unexpected behavior to me. Falling back to a local version as it's currently implemented is a crude form of caching with offline mode support, and at least by me not intended to be used to add custom component jsons. That should be instance level just like with MultiMC, otherwise you'd have two parts to copy around.

I think we can just write some config migration code. e.g. `if meta_url == old_urlL; meta_url = new_url`

anonymous123-code committed:
Address review:
- Add VersionNotFound and IndexNotFound error variants
- move and refactor version_exists
- inline local variables in format usages

anonymous123-code committed:
Use MetaClient instead of Config in Component::into_meta

anonymous123-code committed:
Introduce testing
- fix component_version_exists requiring String instead of &str
- fix get_component_index using the wrong url

kb-1000 reviewed (commented):
I would say the comment about reusing the reqwest and meta client (for the purpose of connection reuse) still stands even now, although to lesser extent now.

Ecorous reviewed (approved):
Looks good to me!

intergrav commented:
hey @anonymous123-code i have no idea where else to contact you so i'll just comment here but i think i accidentally banned you from the discord server - i've unbanned you now, im so sorry i meant to ban the other person

(hopefully you see this i'm not sure where else to ask)
